<!doctype html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport"
          content="width=device-width, user-scalable=no, initial-scale=1.0, maximum-scale=1.0, minimum-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
</head>
<body>

<script src="lib.fx.js"></script>
<!--<script>-->
<!--	// number[]-->
<!--	go(-->
<!--		[1, 2, 3],-->
<!--		L.map(a => Promise.resolve(a + 10)),-->
<!--		take(2),-->
<!--		log-->
<!--	)-->

<!--	// Promise<number>[]-->
<!--	go(-->
<!--		[Promise.resolve(1), Promise.resolve(2), Promise.resolve(3)],-->
<!--		L.map(a => a + 10),-->
<!--		take(2),-->
<!--		log-->
<!--	)-->

<!--	go(-->
<!--		[Promise.resolve(1), Promise.resolve(2), Promise.resolve(3)],-->
<!--		map(a => Promise.resolve(a + 10)),-->
<!--		log-->
<!--	)-->

<!--</script>-->

<script>
	// go(
	// 	[1, 2, 3, 4, 5, 6],
	// 	L.map(a => Promise.resolve(a * a)),
	// 	L.filter(a => {
	// 		console.log(a) // 앞에 filter 를 고쳐서 then 값이 true 일떄 처리해준다
	// 		return a % 2 === 1
	// 	}),
	// 	take(2),
	// 	log
	// )

	go([1, 2, 3, 4, 5, 6],
		L.map(a => Promise.resolve(a * a)), // promise 가 아닌게 와도 계속 nop 을 활용해서 이어나갈수 잇다
		// L.map(a => a * a),
        L.filter(a => {
			// log(a);
			return Promise.resolve(a % 2);
		}),
		L.map(a => {
			// log(a);
			return a * a;
		}),
      take(4),
      log
	)

    /**
     *  중간에 에러 리젝할 경우 그 이후는 무시하게 된다.
     *  * */
	Promise.resolve(1)
	.then(() => Promise.reject('err'))
	.then(_ => console.log('here'))
	.then(_ => console.log('here2'))
	.then(_ => console.log('here3'))
	.catch(e => console.log(e, '결국은 어차피 끝이 중간에 map 들은 무시할꺼야'))
</script>


</body>
</html>